File: Chandan/Searching and Sorting/ExponentialSearch.cpp
// Ref : GeeksForGeeks
// Ref : Wikipedia

// You are Awesome

// Simple program to implement Exponential search over a vector
// Vector must be sorted in ascending order
// O(log(n))

// It can be somewhat compared to jump search
// Rather than dividing the whole list into blocks each of size sqrt(v.size())
// We divide list into 1, 2, 4, 8,.. sized blocks;
// Then similar to jump search, we find block in which value may lie
// and make a binary search into that block

// Returns -1 if value not found
// else return first index of the value found

#include <bits/stdc++.h>
using namespace std;

inline int getLo(int lo, int blocksz, int n){
	return min(lo + blocksz, n);
}

inline int getHi(int lo, int blocksz, int n){
	return min(lo + 2*blocksz, n);
}

inline int getMid(int lo, int hi){
	return (lo + hi)/2;
}

int BinarySearch(std::vector<int> &v, int lo, int hi, int value){
	int mid = getMid(lo, hi);

	while(lo <= hi){
		if(v[mid] == value)		// if value found, return index;
			return mid;

		if(value < v[mid]){		// if value < v[mid], find in right half
			hi = mid - 1;
		}
		else if(v[mid] < value){	// if v[mid] < value, find in left half
			lo = mid + 1;
		}

		mid = getMid(lo, hi);		// set new mid, according as change in start or end
	}

	return -1;		// if value not found, return -1
}

int ExponentialSearch(vector <int> &v, int value){

	int n = v.size()-1;
	int lo = 0;
	int hi = 0;

	if(v[0] == value){
		return 0;
	}

	int blocksz = 1;	// represent current block size;
	while(true){
		lo = getLo(lo, blocksz, n);	// block starting index
		hi = getHi(lo, blocksz, n);	// block ending index

		//cout<<"lo = "<<lo<<", hi = "<<hi<<endl;

		if(v[hi] > value)
			break;

		if(lo == hi) 	// avoid infinite loop due to hi = n;
			return -1;

		blocksz = blocksz << 1;
	}

	// lo --> starting position of block that may contain value
	// hi --> ending position of block that may contain value

	int position = BinarySearch(v, lo, hi, value);	// we can use recursive version of InterpolationSearch

	return position;
}

int main(){
	int n;
	cin>>n;
	vector<int> v(n);
	for(int i=0; i<n; i++) cin>>v[i];

	cout<<"Enter value to search : ";
	int value;
	cin>>value;

	int index = ExponentialSearch(v, value);
	if(index != -1){
		cout<<"Value found at index : "<<index<<endl;
	}
	else{
		cout<<"Value not found"<<endl;
	}

	return 0;
}

File: Rahul/Searching and Sorting/ExponentialSearch.java
// Java program to find an element x in a 
// sorted array using exponential search.
class ExponentialSearch
{
	int binarySearch(int arr[],int ,int ,int);

	// Returns position of first occurenc of
	// x in array
	int exponentialSearch(int arr[],int n,int x)	
	{
	   // If x is present at first location itself
	   if(arr[0]==x)
	     return 0;

	     // Find the range for binary search by
	     // repeated douling
          int i=1;
          while(i<n&&arr[i]<=x)
           i=i*2;

           // Call binary search for found range.
           return binarySearch(arr,i/2,min(i,n),x);

	}
	// A recursive binary search function.It returns
	//location of x in given array arr[l..r] is
	// present otherwise -1
	int binarySearch(int arr[],int l,int r,int x)
	{
	   if(r>=l)
	   {
	       int mid =l+(r-l)/2;

	       //If the element is present at the middle
	       // itself
	       if(arr[mid]==x)
	          return mid;

	       // If element is smaller than mid,then it
	       // can only be present n left subarray
	       if(arr[mid]>x)
	         return binarySearch(arr,l,mid-1,x);

	         //Else the element can only be present
	         // in right subarray
	         return binarySearch(arr,mid+1,r,x);

	   }
	   // We reached here when element is not present
	   // in array
	   return -1;
	}

	// Driver code
	public  static void main(String args[])
	{   
		ExponentialSearch ob=new ExponentialSearch();
	   int arr[]={2,3,4,40,55,5444,66565};
	   int n=arr.length;
	   int x=40;
	   int result = ob.binarySearch(arr,0,n-1,x);
	   if(result==-1)
	   System.out.println("element is not present in array");
	   System.out.printf("element is present at index %d",result);
	   
	}
}